## Series methods and data handling

In [1]:
import pandas as pd

attach and read csv file
* read_csv()
* returns a dataframe

In [2]:
pd.read_csv("https://andybek.com/pandas-drinks")


country  beer_servings  spirit_servings  wine_servings  \
0    Afghanistan            NaN              NaN            NaN   
1        Albania           89.0            132.0           54.0   
2        Algeria           25.0              NaN           14.0   
3        Andorra          245.0            138.0          312.0   
4         Angola          217.0             57.0           45.0   
..           ...            ...              ...            ...   
188    Venezuela          333.0              1.0            3.0   
189      Vietnam          111.0              2.0            1.0   
190        Yemen            6.0              NaN            NaN   
191       Zambia           32.0             19.0            4.0   
192     Zimbabwe           64.0             18.0            4.0   

     total_litres_of_pure_alcohol  
0                             NaN  
1                             4.9  
2                             0.7  
3                            12.4  
4                             5.9  
..                            ...  
188                           7.7  
189                           2.0  
190                           0.1  
191                           2.5  
192                           4.7  

[193 rows x 5 columns]

#### use only necessery columns
* usecols=['column_name', 'another_column']
* specify index_col='name'

In [3]:
alcohol = pd.read_csv("https://andybek.com/pandas-drinks", usecols=['country', 'wine_servings'], index_col='country')
alcohol.head()

wine_servings
country                   
Afghanistan            NaN
Albania               54.0
Algeria               14.0
Andorra              312.0
Angola                45.0

In [4]:
type(alcohol)

pandas.core.frame.DataFrame

#### convert dataframe to a series
* .squeeze()

In [5]:
alcohol_series = pd.read_csv("https://andybek.com/pandas-drinks", usecols=['country', 'wine_servings'], index_col='country').squeeze()
alcohol_series.head()

country
Afghanistan      NaN
Albania         54.0
Algeria         14.0
Andorra        312.0
Angola          45.0
Name: wine_servings, dtype: float64

In [6]:
type(alcohol_series)

pandas.core.series.Series

### series sizing
* size
    - number of elements in the series
* shape
    - tuple of the dimensions for a series, length of the series
* len()
    - python built in function

In [7]:
alcohol_series.size

193

In [8]:
len(alcohol_series)

193

In [9]:
alcohol_series.shape

(193,)

In [10]:
alcohol_series.index

Index(['Afghanistan', 'Albania', 'Algeria', 'Andorra', 'Angola',
       'Antigua & Barbuda', 'Argentina', 'Armenia', 'Australia', 'Austria',
       ...
       'Tanzania', 'USA', 'Uruguay', 'Uzbekistan', 'Vanuatu', 'Venezuela',
       'Vietnam', 'Yemen', 'Zambia', 'Zimbabwe'],
      dtype='object', name='country', length=193)

In [11]:
alcohol_series.values

array([ nan,  54.,  14., 312.,  45.,  45., 221.,  11., 212., 191.,   5.,
        51.,   7.,  nan,  36.,  42., 212.,   8.,  13.,  nan,   8.,   8.,
        35.,  16.,   1.,  94.,   7.,  nan,   7.,  16.,   1.,   4.,   1.,
         1.,   1., 172.,   8.,   3.,   1.,   9.,  74.,  11., 254.,   5.,
       113., 134.,  nan,   1., 278.,   3.,  26.,   9.,   3.,   1.,   2.,
       233.,  nan,  59.,  nan,   1.,  97.,  37.,  59.,   1., 149., 175.,
         1., 218.,  28.,   2.,   2.,  21.,   1.,   1.,   2., 185.,  78.,
        nan,  nan,  nan,  nan, 165.,   9., 237.,   9.,  16.,   1.,  12.,
         2.,   1.,  nan,   6., 123.,  62.,  31.,  nan,   2.,  nan,  56.,
       271.,   4.,   1.,  nan,  nan,   1.,  12.,  nan,  nan,  18.,   5.,
        18.,  nan,   8., 128.,   1.,   5.,  nan,   1.,   8.,  nan,  19.,
       175.,   1.,   1.,   2.,   7., 129.,   1.,  nan,  23.,  18.,   1.,
        74.,  21.,   1.,  56., 339.,   7.,   9.,  18., 167.,  73.,  nan,
        32.,  71.,  11.,  24.,  nan,  14.,  nan,   

#### index and value size should be equal in a series
* check with == operator

In [12]:
alcohol_series.index.size == alcohol_series.values.size

True

#### size and shape should be same in size

In [13]:
alcohol_series.size == alcohol_series.shape[0]

True

## unique values

In [14]:
alcohol_series.is_unique

False

alcohol series does not hold true for unique values, it has some rows similar with other.

In [15]:
alcohol_series.head(5)

country
Afghanistan      NaN
Albania         54.0
Algeria         14.0
Andorra        312.0
Angola          45.0
Name: wine_servings, dtype: float64

### exact number of unique values

In [16]:
alcohol_series.nunique(dropna=False)

72

### monotonicity

In [17]:
pd.Series([1, 2, 3, 4]).is_monotonic_increasing

True

In [18]:
pd.Series([1, 2, 3, 4]).is_monotonic_decreasing

False

In [19]:
pd.Series([1, 4, 3, 9, 2]).is_monotonic_decreasing

False

In [20]:
alcohol_series.is_monotonic_increasing

False

In [21]:
alcohol_series.is_monotonic_decreasing

False

## not null values
* count()

In [22]:
alcohol_series.count()

162

even  if alcohol series has 193 totat rows, there are few null values, and total non-null values are 162. we can verify that using: 
* hasnans

In [23]:
alcohol_series.hasnans

True

count how many null values in the series

In [24]:
sum(alcohol_series.isnull())

31

In [25]:
162 + 31

193

In [26]:
alcohol_series.isnull()

country
Afghanistan     True
Albania        False
Algeria        False
Andorra        False
Angola         False
               ...  
Venezuela      False
Vietnam        False
Yemen           True
Zambia         False
Zimbabwe       False
Name: wine_servings, Length: 193, dtype: bool

## get only null country

In [27]:
alcohol_series[alcohol_series.isnull()].index

Index(['Afghanistan', 'Bangladesh', 'Bhutan', 'Burundi', 'North Korea',
       'Eritrea', 'Ethiopia', 'India', 'Indonesia', 'Iran', 'Iraq', 'Kuwait',
       'Lesotho', 'Libya', 'Malaysia', 'Maldives', 'Marshall Islands',
       'Mauritania', 'Monaco', 'Myanmar', 'Nepal', 'Pakistan', 'Rwanda',
       'San Marino', 'Saudi Arabia', 'Somalia', 'Sri Lanka', 'Sudan',
       'Tajikistan', 'Uganda', 'Yemen'],
      dtype='object', name='country')

In [28]:
list(alcohol_series[alcohol_series.isnull()].index)

['Afghanistan',
 'Bangladesh',
 'Bhutan',
 'Burundi',
 'North Korea',
 'Eritrea',
 'Ethiopia',
 'India',
 'Indonesia',
 'Iran',
 'Iraq',
 'Kuwait',
 'Lesotho',
 'Libya',
 'Malaysia',
 'Maldives',
 'Marshall Islands',
 'Mauritania',
 'Monaco',
 'Myanmar',
 'Nepal',
 'Pakistan',
 'Rwanda',
 'San Marino',
 'Saudi Arabia',
 'Somalia',
 'Sri Lanka',
 'Sudan',
 'Tajikistan',
 'Uganda',
 'Yemen']

In [29]:
len(list(alcohol_series[alcohol_series.isnull()].index))

31

In [30]:
sum(alcohol_series.isna())

31

In [31]:
alcohol_series.isnull().sum()

31

## also availabel in numpy universal function

In [32]:
import numpy as np

In [33]:
alcohol_series[np.isnan].index

Index(['Afghanistan', 'Bangladesh', 'Bhutan', 'Burundi', 'North Korea',
       'Eritrea', 'Ethiopia', 'India', 'Indonesia', 'Iran', 'Iraq', 'Kuwait',
       'Lesotho', 'Libya', 'Malaysia', 'Maldives', 'Marshall Islands',
       'Mauritania', 'Monaco', 'Myanmar', 'Nepal', 'Pakistan', 'Rwanda',
       'San Marino', 'Saudi Arabia', 'Somalia', 'Sri Lanka', 'Sudan',
       'Tajikistan', 'Uganda', 'Yemen'],
      dtype='object', name='country')

In [34]:
len(alcohol_series[np.isnan].index)

31

## everything except null

In [36]:
alcohol_series.notnull()

country
Afghanistan    False
Albania         True
Algeria         True
Andorra         True
Angola          True
               ...  
Venezuela       True
Vietnam         True
Yemen          False
Zambia          True
Zimbabwe        True
Name: wine_servings, Length: 193, dtype: bool

In [37]:
alcohol_series.notnull().sum()

162

In [39]:
alcohol_series.loc[alcohol_series.notnull()]

country
Albania               54.0
Algeria               14.0
Andorra              312.0
Angola                45.0
Antigua & Barbuda     45.0
                     ...  
Vanuatu               11.0
Venezuela              3.0
Vietnam                1.0
Zambia                 4.0
Zimbabwe               4.0
Name: wine_servings, Length: 162, dtype: float64

In [40]:
alcohol_series.count()

162

### Booleans are number
* True = 1
* False = 0

In [41]:
True + 5

6

In [42]:
False + 5

5

## alcohol series 

In [46]:
total_notnull_wine_servings = alcohol_series.count()
total_notnull_wine_servings

162

In [56]:
wine_servings = alcohol_series[alcohol_series.notnull()]
wine_servings.head(5)

country
Albania               54.0
Algeria               14.0
Andorra              312.0
Angola                45.0
Antigua & Barbuda     45.0
Name: wine_servings, dtype: float64

In [49]:
type(wine_servings)

pandas.core.series.Series

In [53]:
total_wine_consumed = sum(wine_servings)
total_wine_consumed

8221.0

In [55]:
less_than_100 = sum(wine_servings[wine_servings < 100])
less_than_100

2416.0

## Drop or fill nans
* dropna()
    - excludes NAs from the series
* fillna()
    - replaces NAs with something else

* both methods return a copy of the series.
* to modify the original series set inplace=True

In [57]:
new_alcohol_series = alcohol_series.dropna()
new_alcohol_series

country
Albania               54.0
Algeria               14.0
Andorra              312.0
Angola                45.0
Antigua & Barbuda     45.0
                     ...  
Vanuatu               11.0
Venezuela              3.0
Vietnam                1.0
Zambia                 4.0
Zimbabwe               4.0
Name: wine_servings, Length: 162, dtype: float64

However, it doesn't overwrite the actual alcohol_series. to overwrite the main dataset we need to pass the condition within the original variable.
* alcohol_series = alcohol_series.dropna()
* or, alcohol_series.dropna(inplace=True)

In [58]:
# fill null with other value that might more meaningfull

In [60]:
fill_null = alcohol_series.fillna(value=100, inplace=False)
fill_null.head(5)

country
Afghanistan    100.0
Albania         54.0
Algeria         14.0
Andorra        312.0
Angola          45.0
Name: wine_servings, dtype: float64